opnode/rollup/derive/payload_attributes.go

```diff
@@ -176,45 +180,151 @@ func UserDeposits(height uint64, receipts []*types.Receipt) ([]*types.DepositTx,
 	return out, nil
 }
 
-type BlockInput interface {
-	ReceiptHash
-	L1Info
-	MixDigest() common.Hash
+func BatchesFromEVMTransactions(config *rollup.Config, txs []*types.Transaction) (out []BatchData) {
+	l1Signer := config.L1Signer()
+	for _, tx := range txs {
+		if to := tx.To(); to != nil && *to == config.BatchInboxAddress {
+			seqDataSubmitter, err := l1Signer.Sender(tx)
+			if err != nil {
+				continue // bad signature, ignore
+			}
+			// some random L1 user might have sent a transaction to our batch inbox, ignore them
+			if seqDataSubmitter != config.BatchSenderAddress {
+				continue // not an authorized batch submitter, ignore
+			}
+			out = append(out, ParseBatches(tx.Data())...)
+		}
+	}
+	return
+}
+
+func FilterBatches(config *rollup.Config, epoch rollup.Epoch, minL2Time uint64, maxL2Time uint64, batches []BatchData) (out []BatchData) {
+	uniqueTime := make(map[uint64]struct{})
+	for _, batch := range batches {
+		if batch.Epoch != epoch {
+			// Batch was tagged for past or future epoch,
+			// i.e. it was included too late or depends on the given L1 block to be processed first.
+			continue
+		}
+		if (batch.Timestamp-config.Genesis.L2Time)%config.BlockTime != 0 {
+			continue // bad timestamp, not a multiple of the block time
+		}
+		if batch.Timestamp < minL2Time {
+			continue // old batch
+		}
+		// limit timestamp upper bound to avoid huge amount of empty blocks
+		if batch.Timestamp >= maxL2Time {
+			continue // too far in future
+		}
+		// Check if we have already seen a batch for this L2 block
+		if _, ok := uniqueTime[batch.Timestamp]; ok {
+			// block already exists, batch is duplicate (first batch persists, others are ignored)
+			continue
+		}
+		uniqueTime[batch.Timestamp] = struct{}{}
+		out = append(out, batch)
+	}
+	return
 }
 
-// PayloadAttributes derives the pre-execution payload from the L1 block info and deposit receipts.
+type L2Info interface {
+	Time() uint64
+}
+
+// PayloadAttributes derives a sequence of pre-execution payload attributes from:
+//  - The L1 information the L1-info deposit is derived from
+//  - The L1 receipts the user deposits are derived from
+//  - The data batches the L2 sequencer work is derived from
+//  - The L2 information of the block the new derived blocks build on
+//
 // This is a pure function.
-func PayloadAttributes(block BlockInput, receipts []*types.Receipt) (*l2.PayloadAttributes, error) {
-	if !CheckReceipts(block, receipts) {
-		return nil, fmt.Errorf("receipts are not consistent with the block's receipts root: %s", block.ReceiptHash())
+func PayloadAttributes(config *rollup.Config, l1Info L1Info, receipts []*types.Receipt, seqWindow []BatchData, l2Info L2Info) ([]*l2.PayloadAttributes, error) {
+	// Retrieve the deposits of this epoch (all deposits from the first block)
+	deposits, err := DeriveDeposits(l1Info, receipts)
```

Comment:
TODO: separate system and user deposits

Reply:
Do you mean pulling out the l1InfoTx deposit?

Reply:
Without adding unique info the the L1 info tx, this working fine but hacky. I think it should probably be cleaned up, but idc about now or in a follow on.